extensions/gitpod-remote/src/ports.ts

```diff
@@ -0,0 +1,122 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Gitpod. All rights reserved.
+ *--------------------------------------------------------------------------------------------*/
+import * as vscode from 'vscode';
+import * as grpc from '@grpc/grpc-js';
+import * as util from 'util';
+import { GitpodExtensionContext, GitpodWorkspacePort } from 'gitpod-shared';
+import { PortsStatus, PortsStatusRequest, PortsStatusResponse } from '@gitpod/supervisor-api-grpc/lib/status_pb';
+import { RetryAutoExposeRequest, TunnelVisiblity } from '@gitpod/supervisor-api-grpc/lib/port_pb';
+
+export async function getSupervisorPorts(context: GitpodExtensionContext) {
+	let supervisorPortList: PortsStatus.AsObject[] = [];
+	try {
+		supervisorPortList = await new Promise<PortsStatus.AsObject[]>((resolve, reject) => {
+			const req = new PortsStatusRequest();
+			const evts = context.supervisor.status.portsStatus(req, context.supervisor.metadata);
+			evts.on('error', reject);
+			evts.on('data', (resp: PortsStatusResponse) => resolve(resp.getPortsList().map(p => p.toObject())));
+		});
+	} catch (e) {
+		context.logger.error('Could not fetch ports info from supervisor', e);
+	}
+	return supervisorPortList;
+}
+
+export function observePortsStatus(context: GitpodExtensionContext): [vscode.EventEmitter<PortsStatus.AsObject[]>, { dispose: () => any }] {
+	const onPortUpdate = new vscode.EventEmitter<PortsStatus.AsObject[]>();
+	let run = true;
+	let stopUpdates: Function | undefined;
+	(async () => {
+		while (run) {
+			try {
+				const req = new PortsStatusRequest();
+				req.setObserve(true);
+				const evts = context.supervisor.status.portsStatus(req, context.supervisor.metadata);
+				stopUpdates = evts.cancel.bind(evts);
+
+				await new Promise((resolve, reject) => {
+					evts.on('end', resolve);
+					evts.on('error', reject);
+					evts.on('data', (resp: PortsStatusResponse) => onPortUpdate.fire(resp.getPortsList().map(p => p.toObject())));
+				});
+			} catch (err) {
+				if (!('code' in err && err.code === grpc.status.CANCELLED)) {
+					context.logger.error('cannot maintain connection to supervisor', err);
+					console.error('cannot maintain connection to supervisor', err);
+				}
+			} finally {
+				stopUpdates = undefined;
+			}
+			await new Promise(resolve => setTimeout(resolve, 1000));
+		}
+	})();
+
+	const toDispose = {
+		dispose() {
+			onPortUpdate.dispose();
+			run = false;
+			if (stopUpdates) {
+				stopUpdates();
+			}
+		}
+	};
+	return [onPortUpdate, toDispose];
+}
+
+export async function tunnelPorts(supervisorPortList: PortsStatus.AsObject[]) {
+	for (const port of supervisorPortList) {
+		if (port.served) {
+			await vscode.env.asExternalUri(vscode.Uri.parse(`http://localhost:${port.localPort}`));
+		}
+	}
+}
+
+interface PortItem { port: GitpodWorkspacePort; isWebview?: boolean }
+
+export function registerPortCommands(context: GitpodExtensionContext) {
```

Comment:
After this PR, we only control ports with new PortsView?

Reply:
It'll be good if we can track if desktop users use our portsView or not, so maybe change `isWebview` field to `source` with values `remote-explorer | ports-view | desktop-ports-view` ?

Reply:
Good catch, I removed the `isWebview` flag for gitpod-remote

> It'll be good if we can track if desktop users use our portsView or not

Analytics already send extension id so we can filter events by gitpod-remote and gitpod-web